File: 069. Sqrt-x/binary-search.py
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        find the last x for which p(x) is false
        """
        if x == 0:
            return 0
        lo = 1
        hi = x
        while lo < hi:
            mid = lo + (hi - lo + 1) // 2
            if p(mid, x):
                hi = mid - 1
            else:
                lo = mid
        if p(lo, x):
            return -1
        return lo
        

def p(n, x):
    """
    predicate that n * n > x
    """
    return n  > (x // n)

File: 069. Sqrt-x/binary-search.cpp
// Time : O(logn)
// Space : O(1)

class Solution {
private:
    // predicate : mid * mid > target?
    bool p(long mid, int target) {
        long square = mid * mid;
        return square > target;
    }
    
public:
    int mySqrt(int x) {
        long lo = 0;
        long hi = x;
        while (lo < hi) {
            long mid = lo + (hi - lo + 1) / 2;
            if (p(mid, x)) {
                hi = mid - 1;
            }
            else {
                lo = mid;
            }
        }
        return lo;
    }
};
